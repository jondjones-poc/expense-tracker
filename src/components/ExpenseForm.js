import React from 'react';
import moment from 'moment';
import { SingleDatePicker } from 'react-dates';

export default class ExpenseForm extends React.Component {
    constructor(props) {
        super(props);
        this.state = {
            description: props.expense ? props.expense.description : '',
            note: props.expense ? props.expense.notes : '',
            amount: props.expense ? (props.expense.amount / 100).toString() : '',
            createdAt: props.expense ? moment(props.expense.createdAt) : moment(),
            calendarFocused: false,
            containsInvalidData: false
        }
        this.onDescriptionChange = this.onDescriptionChange.bind(this);
        this.onNoteChange = this.onNoteChange.bind(this);
        this.onAmountChange = this.onAmountChange.bind(this);
        this.onCalendarDateChange = this.onCalendarDateChange.bind(this);
        this.onCalendarFocusChange = this.onCalendarFocusChange.bind(this);
        this.onSubmit = this.onSubmit.bind(this);
    }
    onDescriptionChange(e) {
        const description = e.target.value;
        this.setState(() => ({description}));
    };
    onNoteChange(e) {
        const note = e.target.value;
        this.setState(() => ({note}));
    };
    onAmountChange(e) {
        const amount = e.target.value;
        if (!amount || amount.match(/^\d{1,}(\.\d{0,2})?$/)) {
            this.setState(() => ({ amount }));
          }
      };
    onCalendarDateChange(createdAt) {
        if (createdAt) {
            this.setState(() => ({createdAt}))
        }
      };
    onCalendarFocusChange(e) {
        this.setState(() => ({calendarFocused: e.focused }))
      };
    onSubmit(e) {
        e.preventDefault();

        if (!this.state.description || !this.state.amount)
        {
            this.setState(() => ({containsInvalidData: true}))
        } else {
            this.setState(() => ({containsInvalidData: false}))
            this.props.onSubmit({
                description: this.state.description,
                amount: parseFloat(this.state.amount,10) * 100,
                createdAt: this.state.createdAt.valueOf(),
                not: this.state.note
            })
        }
    }
    render() {
        return (
                <form onSubmit={this.onSubmit} className="form">
                {this.state.containsInvalidData && <div className="form__error">Description and amount are required</div>}
                    <input
                        type="text"
                        className="text-input"
                        placeholder="Description"
                        autoFocus
                        value={this.state.description}
                        onChange={this.onDescriptionChange}
                    />
                    <input
                        type="number"
                        className="text-input"
                        placeholder="Amount"
                        value={this.state.amount}
                        onChange={this.onAmountChange}
                    />
                    <div className="singledatepicker"  >
                        <SingleDatePicker
                            date={this.state.createdAt}
                            onDateChange={this.onCalendarDateChange}  
                            onFocusChange={this.onCalendarFocusChange} 
                            focused={this.state.calendarFocused}   
                            isOutsideRange={() => false}           
                        />
                    </div>
                    <textarea
                        placeholder="Notes"
                        className="textarea"
                        value={this.state.note}
                        onChange={this.onNoteChange}>
                    </textarea>
                    <div>
                        <button className="button">Save Expense</button>
                    </div>
                </form>
        )
    }
}