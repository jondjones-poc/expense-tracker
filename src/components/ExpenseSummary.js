import numeral from 'numeral';
import React from 'react'
import { connect } from 'react-redux';
import { Link } from 'react-router-dom';
import { adminPageCreateLink } from '../actions/urls';

import getExpenseTotal from '../selectors/expense-total';
import selectExpenses from '../selectors/expenses';

export const ExpenseSummary = ({totalItems, totalAmount}) => {
        const formattedTotalAmount = numeral(totalAmount / 100).format('$0,0.00');
        return (
            <div className="page-header">
                <div className="content-container">
                    <h1 className="page-header__title">
                        <span>You have {totalItems}</span> expense(s) totaling <span>{formattedTotalAmount}</span>
                    </h1>
                    <div className="page-header__actions">
                        <Link to={adminPageCreateLink} className="button" >Create An Expense</Link>
                    </div>
                </div>
            </div>
    )};

const mapStateToProps = (state) => {
    const visibleExpenses = selectExpenses(state.expenses, state.filters);
    const totalAmount = getExpenseTotal(visibleExpenses);

    return {
        totalAmount: totalAmount,
        totalItems: visibleExpenses.length
    }
};

export default connect(mapStateToProps)(ExpenseSummary);