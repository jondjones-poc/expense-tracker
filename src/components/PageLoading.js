import React from 'react';

const PageLoading = () => (
    <div className="loader">
        <img src="/images/loader.gif" className="loader__image"/>
    </div>
)

export default PageLoading