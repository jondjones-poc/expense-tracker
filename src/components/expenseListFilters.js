import React from 'react'
import { connect } from 'react-redux';
import { DateRangePicker } from 'react-dates';
import { setTextFilter, sortByAmount, sortByDate, setStartDate, setEndDate } from '../actions/filterActions';
import moment from 'moment';

export class ExpenseListFilters extends React.Component {
    constructor(props) {
        super(props);
        this.state = {
            calendarFocused: undefined,
        };

        this.onDatesChange = this.onDatesChange.bind(this);
        this.onFocusChange = this.onFocusChange.bind(this);
        this.onTextChange = this.onTextChange.bind(this);
        this.onSortChange = this.onSortChange.bind(this);
    };
    onDatesChange({startDate, endDate}) {
        this.props.setStartDate(startDate);
        this.props.setEndDate(endDate);
    };
    onFocusChange(calendarFocused) {
        this.setState(() => ({calendarFocused}));
    };
    onTextChange(e) {
        this.props.setTextFilter(e.target.value)
    };
    onSortChange(e) {
        if (e.target.value === 'date')
        {
            this.props.sortByDate()
        } else if (e.target.value === 'amount')
        {
            this.props.sortByAmount()
        }
    };
    render() {
        return (
            <div className="content-container">
                <div className="input-group">
                    <div className="input-group__item">
                        <input 
                        type="text" 
                        value={this.props.filters.text} 
                        onChange={this.onTextChange} 
                        className="text-input"
                        placeholder="Search expenses" />
                    </div>
                    <div className="input-group__item">
                        <select
                        className="select"
                        value={this.props.filters.sortBy} 
                        onChange={this.onSortChange}
                    >
                        <option value="date">Date</option>
                        <option value="amount">Amount</option>
                    </select>
                    </div>
                    <div className="input-group__item">
                        <DateRangePicker
                        startDate={this.props.filters.startDate}
                        endDate={this.props.filters.endDate}
                        onDatesChange={this.onDatesChange}
                        focusedInput={this.state.calendarFocused}
                        onFocusChange={this.onFocusChange}
                        startDateId="startDate"
                        endDateId="endDate"
                        isOutsideRange={() => false}
                        showClearDates={true}>
                    </DateRangePicker>
                    </div>
                </div>
            </div>
        )
    };
}

const mapStateToProps = (state) => ({
    filters: state.filters
});

const mapDispatchToProps = (dispatch) => ({
    setTextFilter: (text) => dispatch(setTextFilter(text)),
    sortByDate: () => dispatch(sortByDate()),
    sortByAmount: () => dispatch(sortByAmount()),
    setStartDate: (startDate) => dispatch(setStartDate(startDate)),
    setEndDate: (endDate) => dispatch(setEndDate(endDate))
});


export default connect(mapStateToProps, mapDispatchToProps)(ExpenseListFilters);