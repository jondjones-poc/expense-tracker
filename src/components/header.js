import React from 'react'
import ReactDOM from 'react-dom'
import { Link } from 'react-router-dom';
import { startLogout } from '../actions/authActions';
import { connect } from 'react-redux';

export class Header extends React.Component {
    render() {
        return (
            <header className="header">
                <div className="content-container">
                    <div className="header__content">
                        <Link to={this.props.homePageLink} className="header__title">
                            <h1>
                                Expense Tracker
                            </h1>
                        </Link>
                        <button onClick={this.props.startLogout} className="button button--link">
                            Logout
                        </button>
                    </div>
                </div>
            </header>
        )}
    };

    const mapDispatchToProps = (dispatch) => ({
        startLogout: () => dispatch(startLogout())
    });

    
export default connect(undefined, mapDispatchToProps)(Header);