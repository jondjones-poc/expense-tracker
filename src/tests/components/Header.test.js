import React from 'react';
import { shallow } from 'enzyme';
import { Header } from '../../components/header';

const homePageLink = '/';
const adminPageCreateLink = '/create'
const adminPageEditLink = '/edit/:id';

let wrapper, startLogout;

beforeEach(() => {
    startLogout = jest.fn();
    wrapper = shallow(
        <Header  
         homePageLink={homePageLink}
         adminPageCreateLink={adminPageCreateLink}
         adminPageEditLink={adminPageEditLink}
         startLogout ={startLogout}
     />);
});

test('Header snapshot Tes', () => {
    expect(wrapper).toMatchSnapshot();
});

test('should call logout on button click', () => {
    wrapper.find('button').simulate('click');
    expect(startLogout).toHaveBeenCalled();
});