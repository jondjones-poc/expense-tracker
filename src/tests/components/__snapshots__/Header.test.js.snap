// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Header snapshot Tes 1`] = `
<div>
  <h1>
    Expensify
  </h1>
  <ol>
    <li>
      <NavLink
        activeClassName="is-active"
        ariaCurrent="true"
        exact={true}
        to="/"
      >
        Home
      </NavLink>
    </li>
    <li>
      <NavLink
        activeClassName="is-active"
        ariaCurrent="true"
        to="/create"
      >
        Admin - Create
      </NavLink>
    </li>
    <li>
      <button
        onClick={[MockFunction]}
      >
        Logout
      </button>
    </li>
  </ol>
</div>
`;
